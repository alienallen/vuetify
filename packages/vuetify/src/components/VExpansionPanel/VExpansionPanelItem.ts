// Components
import VExpansionPanel from './VExpansionPanel'
import { VExpandTransition } from '../transitions'
import VIcon from '../VIcon'

// Mixins
import { factory as GroupableFactory } from '../../mixins/groupable'
import Bootable from '../../mixins/bootable'
import Rippleable from '../../mixins/rippleable'

// Utilities
import { keyCodes } from '../../util/helpers'
import mixins, { ExtractVue } from '../../util/mixins'

// Types
import { VNode } from 'vue'

const baseMixins = mixins(
  Bootable,
  GroupableFactory('expansionPanel'),
  Rippleable
)

interface options extends ExtractVue<typeof baseMixins> {
  $refs: {
    header: HTMLElement
  }
  expansionPanel: InstanceType<typeof VExpansionPanel>
}

/* @vue/component */
export default baseMixins.extend<options>().extend({
  name: 'v-expansion-panel-item',

  props: {
    disabled: Boolean,
    readonly: Boolean,
    expandIcon: {
      type: String,
      default: '$vuetify.icons.expand'
    },
    hideActions: Boolean,
    disableIconRotate: Boolean,
    ripple: {
      type: [Boolean, Object],
      default: false
    }
  },

  data: () => ({
    height: 'auto',
    hasMousedown: false
  }),

  computed: {
    classes (): object {
      return {
        'v-expansion-panel-item--active': this.isActive,
        'v-expansion-panel-item--disabled': this.isDisabled,
        'v-expansion-panel-item--has-mousedown': this.hasMousedown,
        'v-expansion-panel-item--disable-item-rotate': this.disableIconRotate,
        ...this.groupClasses
      }
    },
    isDisabled (): boolean {
      return this.expansionPanel.disabled || this.disabled
    },
    isReadonly (): boolean {
      return this.expansionPanel.readonly || this.readonly
    }
  },

  methods: {
    onKeydown (e: KeyboardEvent) {
      // Ensure element is the activeElement
      if (
        e.keyCode === keyCodes.enter &&
        this.$refs.header === document.activeElement
      ) this.click(e)
    },
    click (e: MouseEvent | KeyboardEvent) {
      if (e.detail) this.$refs.header.blur()

      this.$emit('click', e)

      this.isReadonly || this.toggle()
    },
    genBody () {
      return this.$createElement('div', {
        class: 'v-expansion-panel-item__body',
        directives: [{
          name: 'show',
          value: this.isActive
        }]
      }, this.showLazyContent(this.$slots.default))
    },
    genHeader () {
      const children = [...(this.$slots.header || [])]

      if (!this.hideActions) children.push(this.genIcon())

      return this.$createElement('div', {
        staticClass: 'v-expansion-panel-item__header',
        attrs: {
          tabindex: this.isReadonly || this.isDisabled ? null : 0
        },
        directives: [{
          name: 'ripple',
          value: this.ripple
        }],
        on: {
          click: this.click,
          keydown: this.onKeydown,
          mousedown: this.onMousedown,
          mouseup: this.onMouseup
        },
        ref: 'header'
      }, children)
    },
    genIcon () {
      const icon = this.$slots.actions ||
        [this.$createElement(VIcon, this.expandIcon)]

      return this.$createElement('transition', {
        attrs: { name: 'fade-transition' }
      }, [
        this.$createElement('div', {
          staticClass: 'v-expansion-panel-item__header__icon',
          directives: [{
            name: 'show',
            value: !this.isDisabled
          }]
        }, icon)
      ])
    },
    onMousedown (e: Event) {
      this.hasMousedown = true
      this.$emit('mousedown', e)
    },
    onMouseup (e: Event) {
      this.hasMousedown = false
      this.$emit('mouseup', e)
    },
    toggle () {
      this.isBooted = true

      this.$nextTick(() => this.$emit('change'))
    }
  },

  render (h): VNode {
    return h('li', {
      staticClass: 'v-expansion-panel-item',
      class: this.classes,
      attrs: {
        'aria-expanded': String(!!this.isActive)
      }
    }, [
      this.$slots.header && this.genHeader(),
      h(VExpandTransition, [this.genBody()])
    ])
  }
})
