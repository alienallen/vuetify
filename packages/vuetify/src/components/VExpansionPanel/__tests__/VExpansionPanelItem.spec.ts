// Components
import VExpansionPanelItem from '../VExpansionPanelItem'

// Utilities
import {
  mount,
  Wrapper
} from '@vue/test-utils'

describe('VExpansionPanelItem.ts', () => {
  type Instance = InstanceType<typeof VExpansionPanelItem>
  let mountFunction: (options?: object) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(VExpansionPanelItem, {
        slots: {
          header: '<span>foobar</span>'
        },
        provide: {
          expansionPanel: {
            register: () => {},
            unregister: () => {}
          }
        },
        ...options
      })
    }
  })

  it('should work respond to mousedown/up events', () => {
    const wrapper = mountFunction()

    const header = wrapper.find('.v-expansion-panel-item__header')
    expect(wrapper.vm.hasMousedown).toBe(false)

    header.trigger('mousedown')
    expect(wrapper.vm.hasMousedown).toBe(true)

    header.trigger('mouseup')
    expect(wrapper.vm.hasMousedown).toBe(false)
  })

  it('should respond to keyboard.enter', () => {
    const click = jest.fn()
    const wrapper = mountFunction({
      methods: { click }
    })

    const header = wrapper.find('.v-expansion-panel-item__header')

    header.trigger('keydown.up')
    expect(click).not.toHaveBeenCalled()

    header.trigger('keydown.enter')
    expect(click).not.toHaveBeenCalled()

    header.element.focus()
    header.trigger('keydown.enter')
    expect(click).toHaveBeenCalled()
  })

  it('should respond to clicks', () => {
    const click = jest.fn()
    const toggle = jest.fn()
    const wrapper = mountFunction({
      propsData: { readonly: true },
      methods: { toggle }
    })

    wrapper.vm.$on('click', click)
    const spy = jest.spyOn(wrapper.vm.$refs.header, 'blur')
    const header = wrapper.find('.v-expansion-panel-item__header')

    header.trigger('click')

    expect(spy).not.toHaveBeenCalled()
    expect(click).toHaveBeenCalledTimes(1)
    expect(toggle).not.toHaveBeenCalled()

    wrapper.setProps({ readonly: false })

    header.trigger('click')

    expect(spy).not.toHaveBeenCalled()
    expect(click).toHaveBeenCalledTimes(2)
    expect(toggle).toHaveBeenCalledTimes(1)

    // Mock detail
    const event = { detail: 1 }
    wrapper.vm.click(event)

    expect(spy).toHaveBeenCalled()
    expect(click).toHaveBeenCalledTimes(3)
    expect(toggle).toHaveBeenCalledTimes(2)
  })

  // Ensures smooth transition when using the lazy prop
  it('should boot expansion panel item', async () => {
    const change = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        lazy: true
      }
    })

    wrapper.vm.$on('change', change)

    expect(wrapper.vm.isBooted).toBe(false)

    wrapper.vm.toggle()

    expect(wrapper.vm.isBooted).toBe(true)

    await wrapper.vm.$nextTick()

    expect(change).toHaveBeenCalled()
  })

  it('should hide actions and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        hideActions: true
      }
    })

    const snapshot = wrapper.html()
    expect(snapshot).toMatchSnapshot()

    wrapper.setProps({ hideActions: false })

    const snapshot2 = wrapper.html()
    expect(snapshot2).toMatchSnapshot()

    expect(snapshot).not.toEqual(snapshot2)
  })
})
