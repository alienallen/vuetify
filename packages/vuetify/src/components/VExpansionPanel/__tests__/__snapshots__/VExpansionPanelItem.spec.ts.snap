// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VExpansionPanelItem.ts should hide actions and match snapshot 1`] = `

<li aria-expanded="false"
    class="v-expansion-panel-item"
>
  <div tabindex="0"
       class="v-expansion-panel-item__header"
  >
    <span>
      foobar
    </span>
  </div>
  <div class="v-expansion-panel-item__body"
       style="display: none;"
  >
  </div>
</li>

`;

exports[`VExpansionPanelItem.ts should hide actions and match snapshot 2`] = `

<li aria-expanded="false"
    class="v-expansion-panel-item"
>
  <div tabindex="0"
       class="v-expansion-panel-item__header"
  >
    <span>
      foobar
    </span>
    <div class="v-expansion-panel-item__header__icon"
         name="fade-transition"
    >
      <i aria-hidden="true"
         class="v-icon material-icons theme--light"
      >
        $vuetify.icons.expand
      </i>
    </div>
  </div>
  <div class="v-expansion-panel-item__body"
       style="display: none;"
  >
  </div>
</li>

`;
